docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,451 @@
+
+- Provide or install a package management tool
+- Determine which add-ons to install or uninstall on each Cluster
+- Provide a default configuration for each add-on
+- Configure an add-on by resolving or inherit information from the Cluster configuration (i.e. service or pod CIDR for CNI)
+- Install add-ons during cluster creation (immediately after the API server is installed). This operation usually brings in some additional requirements such as: 
+  - Resolve the add-on version for the Cluster’s Kubernetes version
+  - Wait for the add-on to be actually running
+- Provide a mechanism for upgrading an add-on before, during, or after the Cluster upgrade and cleaning up the previous version
+- Support for out-of band upgrades for add-ons not linked to the Cluster lifecycle, i.e. CVE fix
+- Configure pre and post installation steps for each add-on, thus allowing to manage migrations/operations such as:
+  - CPI migration from in-tree to out-of-tree
+  - CoreDNS configuration migration from one version to another
+- Orchestrate add-on deletion before Cluster deletion
+
+This proposal addresses a subset of these problems as part of a first iteration by designing a way for Cluster API to orchestrate an external package management tool for add-on management. An external package management tool is expected to provide all the foundational capabilities for add-on management like add-on packages and repositories, templating/configuration, and management of the creation, upgrade, and deletion workflow etc.
+
+### ClusterAddonProvider Design
+
+When discussing the design of a ClusterAddonProvider, ClusterAddonProvider for Helm will be used as an example of a concrete implementation.
+
+#### ClusterAddonProvider for Helm
+
+ClusterAddonProvider for Helm proposes two new CRDs named HelmChartProxy and HelmReleaseProxy. HelmChartProxy is used to specify a Helm chart with configuration values and a selector identifying the Clusters the add-on applies to. HelmReleaseProxy contains information about the underlying Helm release installed on a cluster from a Helm chart. All the configurations will be done from the HelmChartProxy and the HelmReleaseProxy is simply used to maintain an inventory as well as surface information. Note that the specific fields in both CRDs are subject to change as the project evolves.
+
+This is a link to the prototype for [ClusterAddonProvider for Helm](https://github.com/Jont828/cluster-api-addon-provider-helm), and a demo can be found in this [Cluster API Office Hours recording](https://www.youtube.com/watch?v=dipR1Tzh4jE). The prototype implements HelmChartProxy and HelmReleaseProxy as well as controllers for both CRDs.
+
+#### HelmChartProxy
+
+HelmChartProxy serves as the user interface for defining and configuring Helm charts as well as selecting which workload clusters it will be installed on. 
+
+```go
+// HelmChartProxySpec defines the desired state of HelmChartProxy.
+type HelmChartProxySpec struct {
```

Comment:
Looking at this API type I see a lot of similarities to the [HelmReleases type implemented by Flux](https://fluxcd.io/flux/components/helm/helmreleases/)

To me, that raises a lot of questions:

- Are we reimplementing a solution the community has already come up with?
- How is a user who is already using flux to manage apps supposed to use this in conjunction with Flux?
- If we implemented a ClusterAddonProvider for flux and that would manage helm charts what would that look like?

I'm wondering if AddonProvider should be made even more generic? 

Thank you,

Reply:
Thanks for sharing your concerns. I saw the Flux Helm controller a while back though I'm not super familiar with how it works. To answer your question, the goal of this project is to bring an add-ons user story specifically to Cluster API, not to try to provide a generalized solution. More specifically, we'd like to be able to manage add-ons in CAPI by applying specs to the management Cluster and not having to manually enter the workload cluster and run `helm install`. 

Currently ClusterAddonProvider for Helm does not touch out-of-band Helm releases, meaning if a Helm release was installed from another source besides our controller, it won't be reconciled. So if you already installed Helm charts with Flux this shouldn't touch it. I'd have to look into the Flux controller more to say how we could create ClusterAddonProvider for Flux.

Reply:
> To answer your question, the goal of this project is to bring an add-ons user story specifically to Cluster API, not to try to provide a generalized solution. More specifically, we'd like to be able to manage add-ons in CAPI by applying specs to the management Cluster and not having to manually enter the workload cluster and run helm install.

That's a great goal, but it sounds like this solution is somewhere in between cluster resource sets and full solution for applications that something like Flux or any other GitOps framework provides. In my experience, these are the types of solutions people end up with for application management, even in the case of applying applications to other clusters. 

Has there been any thought into delegating application deployment responsibilities to a solution like this entirely?

Reply:
For example, there's a whole project that is supported, tested, and maintained by the flux project that handles just the Helm bits https://github.com/fluxcd/helm-operator

I'm worried that the `ClusterAddonProvider` controllers will run into the same issues these projects have already encountered.

Reply:
I don't have too much context about the flux project's work with Helm. Could you share a bit about the issues the those projects already encountered? Tagging @fabriziopandini as well.

Reply:
if the flux helm-operator could be tweaked with a minor change to deploy stuff in the target cluster, that seems like a nice approach. Its already a related cncf project.

Reply:
> I don't have too much context about the flux project's work with Helm. Could you share a bit about the issues the those projects already encountered? Tagging @fabriziopandini as well.

To be honest, any bug fix that they've implemented can count as an issue they've encountered. I'm saying this because this implementation is essentially doing what the helm-operator has already accomplished and tested in production clusters. 

Tools like this should be considered in the Alternatives section. An effort should be made to see if we can easily integrate an already existing and well tested component, rather than writing a new one. 

I'm not sure of the timeline to further evaluate tools here, but I can help out flesh out these sections.

Reply:
Flux has several examples of propogating helm-releases to remote clusters including one for [cluster-api that are clearly laid out in the documentation](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api). There's also examples of using [karmada in a multi-cluster environment as well](https://fluxcd.io/flux/use-cases/karmada/). 

At the very least, this needs to be documented in the alternatives section or brought together as a different proposal that can create a way for existing solutions or homegrown ones such as the `HelmProxy`to work with cluster API.

Reply:
> Flux has several examples of propogating helm-releases to remote clusters including one for [cluster-api that are clearly laid out in the documentation](https://fluxcd.io/flux/components/helm/helmreleases/#remote-clusters--cluster-api). There's also examples of using [karmada in a multi-cluster environment as well](https://fluxcd.io/flux/use-cases/karmada/).

Thats a very interesting example, thanks. Bundling the flux helm controller along with cluster-api seems like maybe an easy/viable solution to the problem. It worked well to bundle cert-manager to reduce project load.